app/models/allocation_version.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+class AllocationVersion < ApplicationRecord
+  has_paper_trail
+
+  attr_accessor :responsibility
+
+  ALLOCATE_PRIMARY_POM = 0
+  REALLOCATE_PRIMARY_POM = 1
+  ALLOCATE_SECONDARY_POM = 2
+  REALLOCATE_SECONDARY_POM = 3
+  DEALLOCATE_PRIMARY_POM = 4
+  DEALLOCATE_SECONDARY_POM = 5
+
+  USER = 0
+  OFFENDER_MOVEMENT = 1
+
+  enum event: {
+    allocate_primary_pom: ALLOCATE_PRIMARY_POM,
+    reallocate_primary_pom: REALLOCATE_PRIMARY_POM,
+    allocate_secondary_pom: ALLOCATE_SECONDARY_POM,
+    reallocate_seondary_pom: REALLOCATE_SECONDARY_POM,
+    deallocate_primary_pom: DEALLOCATE_PRIMARY_POM,
+    deallocate_secondary_pom: DEALLOCATE_SECONDARY_POM
+  }
+
+  enum event_trigger: {
+    user: USER,
+    offender_movement: OFFENDER_MOVEMENT
+  }
+
+  scope :allocations, lambda { |nomis_offender_ids|
+    where(nomis_offender_id: nomis_offender_ids)
+  }
+  scope :all_primary_pom_allocations, lambda { |nomis_staff_id|
+    where(
+      primary_pom_nomis_id: nomis_staff_id,
+      event: ALLOCATE_PRIMARY_POM || REALLOCATE_PRIMARY_POM
+    )
+  }
+  scope :active_primary_pom_allocations, lambda { |nomis_staff_id, prison|
+    where(
+      primary_pom_nomis_id: nomis_staff_id,
+      prison: prison,
+      event: ALLOCATE_PRIMARY_POM || REALLOCATE_PRIMARY_POM
+    )
+  }
+  scope :primary_pom_nomis_id, lambda { |nomis_offender_id|
+    allocations(nomis_offender_id).primary_pom_nomis_id
+  }
+
+  def self.deallocate_offender(nomis_offender_id)
+    allocations(nomis_offender_id).
+      update_all(
+        event: DEALLOCATE_PRIMARY_POM,
+        event_trigger: USER
```

Comment:
Is this always a USER trigger? Seems like the type of thing the model might not know about and have to be told.

Reply:
Good spot - Movements can also deallocate an offender ...
I'll fix